feat: add attachments_must_be_private to doctype

marcramser opened the pull request:
## feat: Add "Attachments Must Be Private" setting for DocTypes

This PR introduces a new DocType-level setting called "Attachments Must Be Private" that allows administrators to enforce private attachments for specific DocTypes, providing enhanced security control over sensitive documents and preventing accidental public exposure of confidential files.

**Problem**: Currently, Frappe lacks a mechanism to enforce private attachments for sensitive DocTypes. Users can accidentally make confidential documents publicly accessible, creating potential security risks for organizations handling sensitive data.

**Solution**: This PR adds an "Attachments Must Be Private" checkbox that:
- Automatically enforces private attachments when enabled for a DocType
- Provides server-side validation to prevent security bypasses
- Hides the public/private toggle in the UI when this setting is active
- Maintains full backward compatibility with existing DocTypes and attachments

iamejaaz commented:
Hello,
Could you please share screenshots or videos of the changes, since this is a UI-level change? Also, could you please check the failing test?

marcramser commented:
DocType Edit Screen:
<img width="315" height="189" alt="Image1" src="https://github.com/user-attachments/assets/331f0023-f1cf-44d2-915e-c9bea0e723a9" />

File Upload Screen:
<img width="590" height="242" alt="Image2" src="https://github.com/user-attachments/assets/23297139-376d-4704-9fba-3ff03329c16f" />

marcramser commented:
@iamejaaz Thanks for your feedback!

I've added screenshots and also fixed the Prettier test.
Regarding the remaining test failures. I'm not sure how my PR is causing them. If it is, I’d appreciate any pointers on where they’re breaking and how I might fix them.

marcramser commented:
@iamejaaz any updates on this?

iamejaaz commented:
@marcramser Apologies! I’ll take a look in the next couple of days.

marcramser commented:
@iamejaaz did you have the chance to look into this?

iamejaaz reviewed (changes requested):
Seems like the JSON files were manually modified. Could you please make the changes from the Doctype and then push the changes?

marcramser commented:
@iamejaaz now did the changes in the UI.

iamejaaz commented:
@marcramser 
Could you please modify your commit message to follow the Conventional Commits format?
https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist#pull-request-titles-and-commit-messages

marcramser commented:
I messed up the re name of the commits that destroyed the whole git history. Will see if I have the time to create a new PR for this. Will post the new PR in here once i have it

marcramser committed:
feat: add attachments_must_be_private to doctype

marcramser commented:
@iamejaaz  i now squashed all the commits in one that follows the name guides

marcramser commented:
@iamejaaz any updates?

rmehta commented:
@iamejaaz you are being tagged here, can you please reply? If not, can you at least assign or escalate this to someone else?

iamejaaz reviewed (changes requested):
Really sorry for the delay. Could you please add UI and server test cases as well?
Also, could you please add server-side validation so that when “Attachments Must Be Private” is set, “Make Attachments Public” cannot be enabled at the same time?

marcramser commented:
@iamejaaz and @rmehta

I won’t be continuing work on this PR. The reason is that https://github.com/frappe/frappe/pull/34879
 largely addresses the same problem, even though I still believe this approach would be more sustainable and flexible in the long term.

After this PR being open for almost half a year, with multiple rounds of feedback and long periods without response—while other PRs are merged within a few days or weeks—I’ve become quite discouraged. Given the time and effort already invested, I need to seriously reconsider whether it makes sense to continue contributing to Frappe.

From an outside contributor’s perspective, it also feels quite difficult to get changes merged without already being closely involved with the core team. If that perception is inaccurate, I would really appreciate guidance on how external contributors can better align with the review and decision-making process. For example, is there a Product Owner or someone with a clear vision of the product that contributions can be pitched to before implementation, to ensure they have a chance of being merged? This would help contributors focus their efforts on work that has a realistic path to inclusion.

asieftejani commented:
@marcramser if you watch the youtube (https://youtu.be/ruUfOjEmteY?t=9896) of frappeverse middle east (egypt) 2025
at 2:44:56 @rmehta addressed this issue and said they are developing a platform to monitor contributors in order to promote them at various trust levels
Dont give up hope in frappe yet
